src/memory/redis_store.rs

```diff
@@ -0,0 +1,173 @@
+    for i = 4,(length*2)+3,2
+    do
+        redis.call('SET', ARGV[i], ARGV[i+1])
+    end
+end
+return value
+";
+
+#[derive(Clone)]
+pub struct RedisMemory<Address, Word> {
+    pub manager: ConnectionManager,
+    script_hash: String,
+    a: PhantomData<Address>,
+    w: PhantomData<Word>,
+}
+
+impl<Address, Word> fmt::Debug for RedisMemory<Address, Word> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RedisMemory")
+            .field("connection", &"<redis::Connection>")
+            .finish()
+    }
+}
+
+impl<Address: Sync, Word: Sync> RedisMemory<Address, Word> {
+    /// Connects to a Redis server with a `ConnectionManager`.
+    pub async fn connect_with_manager(manager: ConnectionManager) -> Result<Self, MemoryError> {
+        Ok(Self {
+            manager: manager.clone(),
+            script_hash: redis::cmd("SCRIPT")
+                .arg("LOAD")
+                .arg(GUARDED_WRITE_LUA_SCRIPT)
+                .query_async(&mut manager.clone())
+                .await?,
+            a: PhantomData,
+            w: PhantomData,
+        })
+    }
+
+    /// Connects to a Redis server using the given URL.
+    pub async fn connect(url: &str) -> Result<Self, MemoryError> {
+        let client = redis::Client::open(url)?;
+        let manager = client.get_connection_manager().await?;
+        Self::connect_with_manager(manager).await
+    }
+
+    pub async fn clear_redis_db(&self) -> Result<(), redis::RedisError> {
+        redis::cmd("FLUSHDB")
+            .query_async(&mut self.manager.clone())
+            .await
+    }
+}
+
+impl<const WORD_LENGTH: usize> MemoryADT
+    for RedisMemory<Address<ADDRESS_LENGTH>, [u8; WORD_LENGTH]>
+{
+    type Address = Address<ADDRESS_LENGTH>;
+    type Error = MemoryError;
+    type Word = [u8; WORD_LENGTH];
+
+    async fn batch_read(
+        &self,
+        addresses: Vec<Self::Address>,
+    ) -> Result<Vec<Option<Self::Word>>, Self::Error> {
+        self.manager
+            .clone()
+            .mget(addresses.iter().map(|a| &**a).collect::<Vec<_>>())
+            .await
+            .map_err(Self::Error::from)
+    }
+
+    async fn guarded_write(
+        &self,
+        guard: (Self::Address, Option<Self::Word>),
+        bindings: Vec<(Self::Address, Self::Word)>,
+    ) -> Result<Option<Self::Word>, Self::Error> {
+        let (guard_address, guard_value) = guard;
+        let mut cmd = redis::cmd("EVALSHA");
+        let cmd = cmd
+            .arg(self.script_hash.as_str())
+            .arg(0)
+            .arg(&*guard_address);
+
+        let cmd = if let Some(byte_array) = guard_value {
+            cmd.arg(&byte_array)
+        } else {
+            cmd.arg("false")
+        };
+
+        let cmd = bindings
+            .iter()
+            .fold(cmd.arg(bindings.len()), |cmd, (a, w)| cmd.arg(&**a).arg(w));
+
+        cmd.query_async(&mut self.manager.clone())
+            .await
+            .map_err(std::convert::Into::into)
```

Comment:
Something like that is clearer:
```suggestion
            .map_err(Error::from)
```